### PR TITLE
Update functionality of EventHub*Producer

### PR DESCRIPTION
**Pull request by conniey**
* EventHubAsyncProducer and EventHubProducer can produce events to multiple partitions or round robin depending on SendOptions or BatchOptions.
* Renamed to EventHubAsyncProducerClient and EventHubProducerClient
* Instantiation is through EventHubClientBuilder

**NOTE**: 
* Class documentation and snippets have yet to be updated. I'll do that in another PR after feature work
* I did not rename EventHubClient -> EventHubConnection and make all its methods internal yet. It results in a lot of code churn.
* Afterwards, will be moving all methods like link handling into EventHubConnection, so the link provider is not referenece by EventHubProducer.

Related: #5993

### CODE REVIEW

File: sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventData.java [hunk 1]
```diff
@@ -41,8 +41,8 @@
  * types) and Data section is not supported.
  * </p>
  *
- * @see EventHubProducer
- * @see EventHubAsyncProducer
+ * @see EventHubProducerClient
+ * @see EventHubAsyncProducerClient
```
[the remark this answers is not in the document]
Comment (conniey) [reply sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventData.java]: I'll fix that.

File: sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubAsyncClient.java [hunk 2]
```diff
@@ -63,31 +52,25 @@
     public static final String DEFAULT_CONSUMER_GROUP_NAME = "$Default";
 
     private static final String RECEIVER_ENTITY_PATH_FORMAT = "%s/ConsumerGroups/%s/Partitions/%s";
-    private static final String SENDER_ENTITY_PATH_FORMAT = "%s/Partitions/%s";
 
     private final ClientLogger logger = new ClientLogger(EventHubAsyncClient.class);
     private final MessageSerializer messageSerializer;
-    private final Mono<EventHubConnection> connectionMono;
+    private final EventHubLinkProvider linkProvider;
```
Comment (srnagar) [inline sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubAsyncClient.java]: EventHubAsyncClient should now be package-private.

File: sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubClient.java [hunk 3]
```diff
@@ -40,16 +39,13 @@
 public class EventHubClient implements Closeable {
```
Comment (srnagar) [inline sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubClient.java]: Make this package-private.

File: sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubClientBuilder.java [hunk 4]
```diff
@@ -280,6 +280,34 @@ public EventHubClientBuilder retry(RetryOptions retryOptions) {
...
+     * Creates a new {@link EventHubProducerClient} based on options set on this builder. Every time
+     * {@code buildAsyncProducer()} is invoked, a new instance of {@link EventHubProducerClient} is created.
+     *
+     * @return A new {@link EventHubProducerClient} instance with all the configured options.
+     *
+     * @throws IllegalArgumentException if the credentials have not been set using either {@link
+     * #connectionString(String)} or {@link #credential(String, String, TokenCredential)}. Or, if a proxy is specified
+     * but the transport type is not {@link TransportType#AMQP_WEB_SOCKETS web sockets}.
+     */
+    public EventHubProducerClient buildProducer() {
```
Comment (srnagar) [inline sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubClientBuilder.java]: `buildClient()` and `buildAsyncClient()` should be made package-private too.

File: sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubProducerAsyncClient.java [hunk 5]
```diff
@@ -236,13 +310,18 @@
...
+     *
      * @param events Events to send to the service.
      * @param options The set of options to consider when sending this batch.
      *
      * @return A {@link Mono} that completes when all events are pushed to the service.
      */
     public Mono<Void> send(Iterable<EventData> events, SendOptions options) {
-        Objects.requireNonNull(events, "'options' cannot be null.");
+        if (events == null) {
+            return Mono.error(logger.logExceptionAsError(new NullPointerException("'events' cannot be null.")));
```
Comment (srnagar) [inline sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubProducerAsyncClient.java]: FluxUtils in azure-core has a convenience method for this `monoError(logger, ex)`

File: sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/models/SendOptions.java [hunk 6]
```diff
@@ -54,6 +56,33 @@ public String getPartitionKey() {
...
 
+    /**
+     * Gets the identifier of the Event Hub partition that the {@link EventData events} will be sent to. If the
+     * identifier is not specified, the Event Hubs service will be responsible for routing events that are sent to an
+     * available partition.
+     *
+     * @return The identifier of the Event Hub partition that the {@link EventData events} will be set to. {@code null}
+     *     or an empty string if Event Hubs service is responsible for routing events.
+     */
+    public String getPartitionId() {
```
Comment (srnagar) [inline sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/models/SendOptions.java]: Both `SendOptions` and `BatchOptions` must have a check to either use partitionKey or partition Id and not both.

File: sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubAsyncClient.java [hunk 7]
```diff
@@ -63,31 +52,25 @@
...
     private final EventHubConsumerOptions defaultConsumerOptions;
     private final TracerProvider tracerProvider;
 
     EventHubAsyncClient(ConnectionOptions connectionOptions, TracerProvider tracerProvider,
-                        MessageSerializer messageSerializer, Mono<EventHubConnection> eventHubConnectionMono) {
+                        MessageSerializer messageSerializer, EventHubLinkProvider linkProvider) {
 
         this.connectionOptions = Objects.requireNonNull(connectionOptions, "'connectionOptions' cannot be null.");
         this.tracerProvider = Objects.requireNonNull(tracerProvider, "'tracerProvider' cannot be null.");
         this.eventHubName = connectionOptions.getEntityPath();
```
Comment (srnagar) [inline sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubAsyncClient.java]: This client should also have a method to retrieve `fullyQualifiedNamespace`

File: sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubClient.java (hunk 3 above)
Comment (conniey) [reply sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubClient.java]: That's in another PR. it breaks a lot of things. After this is merged, I'll create another that makes it package private and updates it.

File: sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/models/SendOptions.java (hunk 6 above)
Comment (conniey) [reply sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/models/SendOptions.java]: The validation happens when you go to send the message through AsyncProducer.send(). I wanted to delay validation until they go to invoke those.

File: sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubAsyncClient.java (hunk 2 above)
Comment (conniey) [reply sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubAsyncClient.java]: This is going to be addressed in another PR with the rename of EventHubClient -> EventHubConnection (and all its methods package private).

File: sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubClientBuilder.java (hunk 4 above)
Comment (conniey) [reply sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubClientBuilder.java]: This is going to be addressed in another PR with the rename of EventHubClient -> EventHubConnection (and all its methods package private).

File: sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubAsyncClient.java (hunk 7 above)
Comment (conniey) [reply sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubAsyncClient.java]: Fixed

**Pull request merged**